include/s3select_json_parser.h

```diff
@@ -0,0 +1,470 @@
+*/
+class ChunksStreamer : public rapidjson::MemoryStream {
+
+  //purpose: adding a method `resetBuffer` that enables to parse chunk after chunk
+  //per each new chunk it reset internal data members
+  public:
+
+    std::string internal_buffer;
+    const Ch* next_src_;
+    size_t next_size_;
+
+    ChunksStreamer():rapidjson::MemoryStream(0,0){next_src_=0;next_size_=0;}
+
+    ChunksStreamer(const Ch *src, size_t size) : rapidjson::MemoryStream(src,size){next_src_=0;next_size_=0;}
+
+    //override Peek methode
+    Ch Peek() //const 
+    {
+      if(RAPIDJSON_UNLIKELY(src_ == end_))
+      {
+	if(next_src_)//next chunk exist
+	{//upon reaching to end of current buffer, to switch with next one
+	  src_ = next_src_;
+	  begin_ = src_;
+	  size_ =next_size_;
+	  end_ = src_ + size_;
+
+	  next_src_ = 0;
+	  next_size_ = 0;
+	  return *src_;
+	}
+	else return 0;
+      }
+      return *src_;
+    }
+
+    //override Take method
+    Ch Take() 
+    {
+      if(RAPIDJSON_UNLIKELY(src_ == end_))
+      {
+	if(next_src_)//next chunk exist
+	{//upon reaching to end of current buffer, to switch with next one
+	  src_ = next_src_;
+	  begin_ = src_;
+	  size_ = next_size_;
+	  end_ = src_ + size_;
+
+	  next_src_ = 0;
+	  next_size_ = 0;
+	  return *src_;
+	}
+	else return 0;
+      }
+      return *src_++;
+    }
+
+    void resetBuffer(char* buff, size_t size)
+    {
+      if(!src_)
+      {//first time calling
+	begin_ = buff;
+	src_ = buff;
+	size_ = size;
+	end_= src_ + size_;
+	return;
+      }
+
+      if(!next_src_)
+      {//save the next-chunk that will be used upon parser reaches end of current buffer
+	next_src_ = buff;
+	next_size_ = size;
+      }
+      else
+      {// should not happen
+	std::cout << "can not replace pointers!!!" << std::endl;//TODO exception
+	return;
+      }
+    }
+
+    void saveRemainingBytes()
+    {//this routine called per each new chunk
+      //savine the remaining bytes, before its overriden by the next-chunk.
+      size_t copy_left_sz = getBytesLeft(); //should be very small
+      internal_buffer.assign(src_,copy_left_sz);
+      
+      src_ = internal_buffer.data();
+      begin_ = src_;
+      size_ = copy_left_sz;
+      end_= src_ + copy_left_sz;
+    }
+
+    size_t getBytesLeft() { return end_ - src_; }
+
+};
+
+class JsonParserHandler : public rapidjson::BaseReaderHandler<rapidjson::UTF8<>, JsonParserHandler> {
+
+  public:
+
+    typedef enum {OBJECT_STATE,ARRAY_STATE} en_json_elm_state_t;
+
+    typedef std::pair<std::vector<std::string>, s3selectEngine::value> json_key_value_t;
+
+    enum class row_state
+    {
+      NA,
+      OBJECT_START_ROW,
+      ARRAY_START_ROW,
+      END_ROW
+    };
+
+    row_state state = row_state::NA;
+// TODO its possible there is no need to std::vector<std::string>, the json-idx is enough
+    std::function <int(JsonParserHandler::json_key_value_t&,int)> m_exact_match_cb;
+
+    std::vector <std::vector<std::string>> query_matrix{};
+    int row_count{};
```

Comment:
the row-count should be aligned with the SQL processing.

upon a complete row accepted by the parser, the SQL engine starts sql-processing.
the row-count should be align with that